File: Piko/Sources/SFBufferOperator.m
//
//  SFBufferOperator.m
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

#import "SFBufferOperator.h"

#import "SFBufferStorage.h"

@implementation SFBufferOperator

- (instancetype)initWithStorage:(SFBufferStorage *)storage
{
    self = [super init];
    if (self != nil) {
        _storage = storage;
        _buffer = [storage pull];
    }
    return self;
}

- (void)dealloc
{
    [_storage push:_buffer];
}

@end

File: Piko/Sources/SFBufferOperator.swift
//
//  SFBufferOperator.swift
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

import Foundation

@objc class SFBufferOperator: NSObject {

    let buffer: SFBuffer
    let storage: SFBufferStorage

    init(storage: SFBufferStorage) {
        self.storage = storage
        self.buffer = storage.pull()
        super.init()
    }

    deinit {
        storage.push(buffer)
    }
}
